precompiles/dispatch-lockdrop/src/lib.rs

```diff
@@ -0,0 +1,245 @@
+use sp_core::{crypto::AccountId32, H160, H256};
+use sp_core::{ecdsa, U256};
+use sp_io::hashing::keccak_256;
+use sp_runtime::traits::Zero;
+use sp_std::vec::Vec;
+
+#[cfg(test)]
+mod mock;
+#[cfg(test)]
+mod tests;
+
+pub const LOG_TARGET: &str = "dispatch-lockdrop-precompile";
+
+// ECDSA signature bytes
+type ECDSASignatureBytes = ConstU32<65>;
+
+// `DecodeLimit` specifies the max depth a call can use when decoding, as unbounded depth
+// can be used to overflow the stack.
+// Default value is 8, which is the same as in XCM call decoding.
+pub struct DispatchLockdrop<Runtime, DispatchValidator = (), DecodeLimit = ConstU32<8>>(
+    PhantomData<(Runtime, DispatchValidator, DecodeLimit)>,
+);
+
+#[precompile_utils::precompile]
+impl<Runtime, DispatchValidator, DecodeLimit>
+    DispatchLockdrop<Runtime, DispatchValidator, DecodeLimit>
+where
+    Runtime: pallet_evm::Config + pallet_unified_accounts::Config,
+    <Runtime::RuntimeCall as Dispatchable>::RuntimeOrigin: From<Option<Runtime::AccountId>>,
+    Runtime::RuntimeCall: Dispatchable<PostInfo = PostDispatchInfo> + GetDispatchInfo,
+    <Runtime as Config>::AccountId: IsType<AccountId32>,
+    <Runtime as Config>::AccountId: From<[u8; 32]>,
+    DispatchValidator:
+        DispatchValidateT<<Runtime as Config>::AccountId, <Runtime as Config>::RuntimeCall>,
+    DecodeLimit: Get<u32>,
+{
+    #[precompile::public("dispatch_lockdrop_call(bytes,bytes32,bytes)")]
+    fn dispatch_lockdrop_call(
+        handle: &mut impl PrecompileHandle,
+        call: UnboundedBytes,
+        account_id: H256,
+        signature: BoundedBytes<ECDSASignatureBytes>,
+    ) -> EvmResult<bool> {
+        log::trace!(
+            target: LOG_TARGET,
+            "raw arguments: call: {:?}, account_id: {:?}, signature: {:?}",
+            call,
+            account_id,
+            signature
+        );
+
+        let target_gas = handle.gas_limit();
+        let caller: H160 = handle.context().caller.into();
+        let input: Vec<u8> = call.into();
+        let signature_bytes: Vec<u8> = signature.into();
+        let account_id = AccountId32::new(account_id.into()).into();
+
+        // 1. Decode the call
+        let call =
+            match Runtime::RuntimeCall::decode_with_depth_limit(DecodeLimit::get(), &mut &*input) {
+                Ok(c) => c,
+                Err(_) => {
+                    log::trace!(target: LOG_TARGET, "Error: could not decode call");
+                    return Ok(false);
+                }
+            };
+
+        // 2. Check that gas limit is not exceeded
+        let info = call.get_dispatch_info();
+        if let Some(gas) = target_gas {
+            let valid_weight = info.weight.ref_time()
+                <= Runtime::GasWeightMapping::gas_to_weight(gas, false).ref_time();
+            if !valid_weight {
+                log::trace!(target: LOG_TARGET, "Error: gas limit exceeded");
+                return Ok(false);
+            }
+        }
```

Comment:
Just `.record_external_cost(...)` should be good enough here?

It will throw an error in case the result would be out of gas.

Reply:
You should account for `record_external_cost` and `record_cost`